lib/executor/ramping_vus.go

```diff
@@ -486,174 +486,235 @@ func (vlvc RampingVUsConfig) HasWork(et *lib.ExecutionTuple) bool {
-	progressFn := func() (float64, []string) {
-		spent := time.Since(startTime)
-		currentlyActiveVUs := atomic.LoadInt64(activeVUsCount)
-		vus := fmt.Sprintf(vusFmt+"/"+vusFmt+" VUs", currentlyActiveVUs, maxVUs)
-		if spent > regularDuration {
-			return 1, []string{vus, regularDuration.String()}
-		}
-		progVUs := fmt.Sprintf(vusFmt+"/"+vusFmt+" VUs", currentlyActiveVUs, maxVUs)
-		progDur := pb.GetFixedLengthDuration(spent, regularDuration) + "/" + regularDurationStr
-		return float64(spent) / float64(regularDuration), []string{progVUs, progDur}
-	}
-	vlv.progress.Modify(pb.WithProgress(progressFn))
-	go trackProgress(parentCtx, maxDurationCtx, regDurationCtx, vlv, progressFn)
-
-	// Actually schedule the VUs and iterations, likely the most complicated
-	// executor among all of them...
-	runIteration := getIterationRunner(vlv.executionState, vlv.logger)
-	getVU := func() (lib.InitializedVU, error) {
-		initVU, err := vlv.executionState.GetPlannedVU(vlv.logger, false)
-		if err != nil {
-			vlv.logger.WithError(err).Error("Cannot get a VU from the buffer")
-			cancel()
-		} else {
-			activeVUs.Add(1)
-			atomic.AddInt64(activeVUsCount, 1)
-			vlv.executionState.ModCurrentlyActiveVUsCount(+1)
-		}
-		return initVU, err
-	}
-	returnVU := func(initVU lib.InitializedVU) {
-		vlv.executionState.ReturnVU(initVU, false)
-		atomic.AddInt64(activeVUsCount, -1)
-		activeVUs.Done()
-		vlv.executionState.ModCurrentlyActiveVUsCount(-1)
+		"type":      vlv.config.GetType(),
+		"startVUs":  vlv.config.GetStartVUs(vlv.executionState.ExecutionTuple),
+		"maxVUs":    maxVUs,
+		"duration":  regularDuration,
+		"numStages": len(vlv.config.Stages),
+	}).Debug("Starting executor run...")
+
+	runState := &rampingVUsRunState{
+		executor:       vlv,
+		wg:             new(sync.WaitGroup),
+		vuHandles:      make([]*vuHandle, maxVUs),
+		maxVUs:         maxVUs,
+		activeVUsCount: new(int64),
+		started:        startTime,
+		runIteration:   getIterationRunner(vlv.executionState, vlv.logger),
 	}
 
+	progressFn := runState.makeProgressFn(regularDuration)
 	maxDurationCtx = lib.WithScenarioState(maxDurationCtx, &lib.ScenarioState{
 		Name:       vlv.config.Name,
 		Executor:   vlv.config.Type,
-		StartTime:  startTime,
+		StartTime:  runState.started,
 		ProgressFn: progressFn,
 	})
+	vlv.progress.Modify(pb.WithProgress(progressFn))
+	go trackProgress(ctx, maxDurationCtx, regularDurationCtx, vlv, progressFn)
 
-	vuHandles := make([]*vuHandle, maxVUs)
-	for i := uint64(0); i < maxVUs; i++ {
-		vuHandle := newStoppedVUHandle(
-			maxDurationCtx, getVU, returnVU, vlv.nextIterationCounters,
-			&vlv.config.BaseConfig, vlv.logger.WithField("vuNum", i))
-		go vuHandle.runLoopsIfPossible(runIteration)
-		vuHandles[i] = vuHandle
+	defer runState.wg.Wait()
+	runState.populateVUHandles(maxDurationCtx, cancel)
+	for i := uint64(0); i < runState.maxVUs; i++ {
+		go runState.vuHandles[i].runLoopsIfPossible(runState.runIteration)
```

Comment:
Here 1 loop turned into 2 loops after refactoring :joy: Is it possible to put all the `vuHandle` initialization into one `runState`'s method with 1 loop?

Reply:
@yorugac

Some of this was because of a previous review request.

I like to keep it this way. Making the loop in the `Run` method here allows us to see all the goroutines we're launching and easier to see what's going on. You can see that the `Run` method launches all the necessary goroutines, and we can read it like a script.

Reply:
Looking at this now I also feel that this will likely be better to just be part of `populateVUHandles` - it already hides a bunch of stuff, you can just rename it to `loopVUHandlesIfPossible` ;) 

What was the original concern as I can't find it?

Reply:
`populateVUHandles` only does what it says, and it does a single thing: populating VU handles. The `Run` method is responsible for orchestrating them.

What do you mean by the original concern?

Reply:
> What do you mean by the original concern? 

you wrote "Some of this was because of previous review requests." so I am asking what was the concern in the original review requests as I can't find them.

Reply:
Ah, got it now, thanks. Multiple people have reviewed this PR many many times, so I don't remember, sorry :(

Reply:
I couldn't find mention of it in my email history either.

I still think that it should be 1 loop / 1 method :) I.e. the readability won't be that much worse while adding loops like this makes at least me automatically wonder about "what if" of potential performance losses. How about adding a comment in `Run` mentioning that this call starts goroutines?

Reply:
Here you go 😅 95fb52d5c1639014c77a028403f5152513ebff8e